### PR TITLE
Implement Server-Side Apply mode

### PR DESCRIPTION
**Pull request by lblackstone**
### Proposed changes
Add a mode to the provider that uses Server-Side Apply (SSA) for resource management. See https://github.com/pulumi/pulumi-kubernetes/issues/2011 for additional details on the background and design of this feature.

### Related issues (optional)
Fix https://github.com/pulumi/pulumi-kubernetes/issues/2016

### CODE REVIEW

File: provider/pkg/ssa/actions.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// specified field manager. Any fields that no longer have a manager will return to a default value, so this isn't
+// guaranteed to return the resource back to the same state that it was in before the patch.
+//
+// This operation will fail unless another manager is responsible for a valid resource spec. In other words, if the
+// specified field manager is the sole manager of a resource, then this will fail with a validation error for most
+// resources.
+func Relinquish(
+	ctx context.Context,
+	client dynamic.ResourceInterface,
+	inputs *unstructured.Unstructured,
```
Comment (viveklak) [inline provider/pkg/ssa/actions.go:37]: Passing the entire inputs was pretty confusing for me and I initially parsed this to suggest we were using the inputs as the body for the patch (which wouldn't do the relinquishing at all). Could you replace this with just the name so it makes it explicit how its being used?

**Review by viveklak — CHANGES_REQUESTED** [review]
Looking great! Some minor suggestions and questions but its coming together nicely!

File: provider/pkg/provider/provider.go [hunk 2]
```diff
@@ -2672,7 +2785,28 @@ func initialAPIVersion(state resource.PropertyMap, oldConfig *unstructured.Unstr
...
 }
 
-func checkpointObject(inputs, live *unstructured.Unstructured, fromInputs resource.PropertyMap, initialAPIVersion string) resource.PropertyMap {
+// fieldManagerName returns the name to use for the Server-Side Apply fieldManager. The values are looked up with the
+// following precedence:
+// 1. Resource annotation (this will likely change to a typed option field in the next major release)
+// 2. Value from the Pulumi state
+// 3. Randomly generated name
+func fieldManagerName(state resource.PropertyMap, inputs *unstructured.Unstructured) string {
+	if v := metadata.GetAnnotationValue(inputs, metadata.AnnotationPatchManager); len(v) > 0 {
```
[the remark this answers is not in the document]
Comment (viveklak) [reply provider/pkg/provider/provider.go:2794]: Discussed online and this is to support an explicit ownership transfer similar to https://kubernetes.io/docs/reference/using-api/server-side-apply/#transferring-ownership

File: provider/pkg/await/await.go [hunk 3]
```diff
@@ -368,9 +394,6 @@ func Update(c UpdateConfig) (*unstructured.Unstructured, error) {
 
 	var currentOutputs *unstructured.Unstructured
 	if clients.IsCRD(c.Inputs) {
-		// Note: This feature is currently enabled with a provider feature flag, but is expected to eventually become
-		// the default behavior.
-
 		// CRDs require special handling to update. Rather than computing a patch, replace the CRD with a PUT
 		// operation (equivalent to running `kubectl replace`). This is accomplished by getting the `resourceVersion`
 		// of the existing CRD, setting that as the `resourceVersion` in the request, and then running an update. This
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/await/await.go:399]: I'm pretty sure it still requires special handling. Could be interesting to try as a followup, but I did my best to leave existing code paths alone for this PR.

File: provider/pkg/await/await.go [hunk 4]
```diff
@@ -177,16 +178,41 @@ func Creation(c CreateConfig) (*unstructured.Unstructured, error) {
...
-					return &namespaceError{c.Inputs}
+			if c.ServerSideApply {
+				force := metadata.IsAnnotationTrue(c.Inputs, metadata.AnnotationPatchForce)
+				options := metav1.PatchOptions{
+					FieldManager: c.FieldManager,
+					Force:        &force,
+				}
+				if c.DryRun {
+					options.DryRun = []string{metav1.DryRunAll}
+				}
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/await/await.go:189]: This option is confusingly named, and actually refers to Pulumi running in preview mode vs a regular update.

File: provider/pkg/metadata/annotations.go [hunk 5]
```diff
@@ -33,6 +33,9 @@ const (
 	AnnotationInitialAPIVersion = AnnotationPrefix + "initialApiVersion"
 	AnnotationReplaceUnready    = AnnotationPrefix + "replaceUnready"
 
+	AnnotationPatchForce   = AnnotationPrefix + "patchForce"
+	AnnotationPatchManager = AnnotationPrefix + "patchManager"
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/metadata/annotations.go:37]: I could do `patchFieldManager`, but was intentionally naming all Patch options with the `patch` prefix.

File: provider/pkg/provider/provider.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/provider/provider.go:2794]: This is to support the [ownership transfer](https://kubernetes.io/docs/reference/using-api/server-side-apply/#transferring-ownership) use case.

File: provider/pkg/provider/provider.go [hunk 6]
```diff
@@ -2260,6 +2298,21 @@ func (k *kubeProvider) Update(
 			nil)
 	}
 
+	if k.serverSideApplyMode {
+		// For non-preview updates, drop the old fieldManager if the value changes.
+		if !req.GetPreview() && fieldManagerOld != fieldManager {
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/provider/provider.go:2303]: This is part of the [ownership transfer](https://kubernetes.io/docs/reference/using-api/server-side-apply/#transferring-ownership) support. This code is triggered if a user changes the value of the annotation.

File: provider/pkg/ssa/actions.go (hunk 1 above)
Comment (lblackstone) [reply provider/pkg/ssa/actions.go:37]: We chatted about this separately, but I realized that I'm also using the inputs to populate the minimal spec (GVK + namespace/name). It's simpler to leave as is, but I can clarify the usage with a comment.

File: provider/pkg/await/await.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/await/await.go:189]: Opened https://github.com/pulumi/pulumi-kubernetes/pull/2040 to fix this

File: provider/pkg/await/await.go [hunk 7]
```diff
@@ -177,16 +178,41 @@ func Creation(c CreateConfig) (*unstructured.Unstructured, error) {
...
+				objYAML, err := yaml.Marshal(c.Inputs.Object)
+				if err != nil {
+					return err
+				}
+				outputs, err = client.Patch(
+					context.TODO(), c.Inputs.GetName(), types.ApplyPatchType, objYAML, options)
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/await/await.go:195]: Fixed in https://github.com/pulumi/pulumi-kubernetes/pull/2042

File: provider/pkg/provider/provider.go [hunk 8]
```diff
@@ -2546,43 +2652,50 @@ func (k *kubeProvider) isDryRunDisabledError(err error) bool {
...
+		return nil, nil, false, err
 	}
 
 	// The server-side patch succeeded.
-	return ssPatch, ssPatchBase, true
+	return ssPatch, ssPatchBase, true, nil
 }
 
 func (k *kubeProvider) withLastAppliedConfig(config *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	if k.serverSideApplyMode {
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/provider/provider.go:2695]: Nevermind. Fixed

File: provider/pkg/provider/provider.go [hunk 9]
```diff
@@ -2546,43 +2652,50 @@ func (k *kubeProvider) isDryRunDisabledError(err error) bool {
 }
 
 // tryServerSidePatch attempts to compute a server-side patch. Returns true iff the operation succeeded.
-func (k *kubeProvider) tryServerSidePatch(oldInputs, newInputs *unstructured.Unstructured, gvk schema.GroupVersionKind,
-) ([]byte, map[string]interface{}, bool) {
+func (k *kubeProvider) tryServerSidePatch(
+	oldInputs, newInputs *unstructured.Unstructured,
+	gvk schema.GroupVersionKind,
+	fieldManager string,
+) (ssPatch []byte, ssPatchBase map[string]interface{}, ok bool, err error) {
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply provider/pkg/provider/provider.go:2659]: I was getting confused about what the return values were, so I'll leave as is for now.

File: CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,7 @@
 ## Unreleased
 
+- Implement Server-Side Apply mode (https://github.com/pulumi/pulumi-kubernetes/pull/2029)
```
Comment (viveklak) [inline CHANGELOG.md:3]: Maybe add some more details on how to enable this. (And mark it preview etc.)

**Review by viveklak — APPROVED** [review]

File: CHANGELOG.md (hunk 10 above)
Comment (lblackstone) [reply CHANGELOG.md:3]: I'll add that in a followup PR. This isn't really feature-complete yet, so I'm not quite ready to encourage usage.

**Pull request merged**